url2df0f7c811_przemekgalazka_legacycode_tgz-pJ8-pl_cc_core_cmd_CommandJ8/src/ccproxy/AgentConnection.java

```diff
@@ -0,0 +1,1042 @@
+				out.close();
+				in.close();
+			} else {
+				cleanup(null, mainThread, true);
+			}
+			if (!loggedout)
+				log.error(e);
+		}
+	}
+
+	private void stopPingThread() {
+		log.info("Stopping ping thread");
+        if (agentConnectionPing!=null) agentConnectionPing.stopThread();
+		
+	}
+
+	private void cleanup(Thread pingThread, Thread mainThread,
+			boolean raisedFromMainThread) {
+
+		p.log.info("Prepare to clean after connection: Agent: " + agentNumber
+				+ ", exten: " + agentExten + ", IP: " + ip + ":" + port);
+		if (loggedout)
+			return;
+		loggedout = true;
+
+		if (raisedFromMainThread) {
+			p.log.info("Cleanup raised from main thread: Agent: " + agentNumber
+					+ ", exten: " + agentExten + ", IP: " + ip + ":" + port);
+			if (agentConnectionPing!=null) agentConnectionPing.stopThread();
+		} else {
+			p.log.info("Cleanup raised from ping thread: Agent: " + agentNumber
+					+ ", exten: " + agentExten + ", IP: " + ip + ":" + port);
+			mainThread.interrupt();
+		}
+
+		p.log.info("Try to close connection: Agent: " + agentNumber
+				+ ", exten: " + agentExten + ", IP: " + ip + ":" + port);
+		
+		out.close();
+		in.close();
+
+		p.log.info("Agent connection closed: Agent: " + agentNumber
+				+ ", exten: " + agentExten + ", IP: " + ip + ":" + port);
+		//nie wysyłamy pauzy przed wylogowaniem 
+		//p.gadajAsterisk.sendMessage(this, "pause"); 
+		if (clearexit) {
+			p.log.info("Clear exit received. Logout agent: Agent: "
+					+ agentNumber + ", exten: " + agentExten + ", IP: " + ip
+					+ ":" + port);
+			p.gadajAsterisk.sendMessage(this, "logout");
+		} else {
+			p.log.info("Unclear exit received. Logout agentt: Agent: "
+					+ agentNumber + ", exten: " + agentExten + ", IP: " + ip
+					+ ":" + port);
+			p.gadajAsterisk.sendMessage(this, "logout");
+		}
+		p.asteriskQueueList.removeAgentConnection(this);
+		p.agenci.remove(this);
+		AsteriskCmd cmd = new AsteriskCmd(p);
+
+		GadajAsterisk.out.print(cmd.delAgentExten(agentExten));
+		GadajAsterisk.out.print(cmd.delAgentNumber(agentNumber));
+
+		try {
+			in.close();
+		} catch (Exception e) {
+			p.log.info("Wywalono in: " + e.getMessage());
+		}
+		try {
+			out.close();
+		} catch (Exception e) {
+			p.log.info("Wywalono out: " + e.getMessage());
+		}
+		System.out.println("Agent connection closed.");
+
+		;
+
+	}
+
+	private void agentMakeQueuePenalty(String agentNumber) {
+		String qname, penalty;
+		ResultSet rs;
+		
+			rs = p.dbConn
+					.query("select nazwa, penalty from v_agent_queue where numer='"
+							+ agentNumber + "';");
+			try {
+				while (rs.next()) {
+					qname = new String(rs.getString("nazwa"));
+					penalty = new String(rs.getString("penalty"));
+					p.log.info("Agent: " + agentNumber + ", Queue: " + qname
+							+ ", penalty: " + penalty);
+					this.queuePenalty.put(qname, penalty);
+				}
+			} catch (SQLException e) {
+				e.printStackTrace();
+			}
+	}
+
+	private boolean agentAuthorization() throws IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 73. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 775-876 into a separate method.